File: Grind75/33-EvaluateReversePolish/python/EvaluateReversePolish.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from typing import List, Deque
from collections import deque
import operator

class Solution(object):
  def evalRPN(self, tokens: List[str]) -> int:
    stack: Deque[int] = deque()
    for token in tokens:
      match token:
        case "+":
          n2, n1 = stack.pop(), stack.pop()
          stack.append(n1 + n2)
        case "-":
          n2, n1 = stack.pop(), stack.pop()
          stack.append(n1 - n2)
        case "*":
          n2, n1 = stack.pop(), stack.pop()
          stack.append(n1 * n2)
        case "/":
          n2, n1 = stack.pop(), stack.pop()
          stack.append(int(n1 / n2))
        case "_":
          stack.append(int(token))
    return stack[-1]

  def evalRPN_fancy(self, tokens: List[str]) -> int:
    stack: Deque[int] = deque()
    ops = {
      "+": operator.add, "-": operator.sub,
      "*": operator.mul, "/": operator.truediv
    }
    for token in tokens:
      match token:
        case "+" | "-" | "*" | "/":
          n2, n1 = stack.pop(), stack.pop()
          stack.append(int(ops[token](n1, n2)))
        case _:
          stack.append(int(token))
    return stack.pop()

if __name__ == '__main__':
  soln = Solution()
  print(soln.evalRPN_fancy(["10","6","+"]))

File: Grind75/33-EvaluateReversePolish/cpp/EvaluateReversePolish.cpp
#include "EvaluateReversePolish.h"
#include <vector>
#include <string>
#include <stack>

using namespace std;

int EvaluateReversePolish::evalRPN(vector<string>& tokens) {
  stack<int> stack;

  for (string token : tokens) {
    if (token == "+") {
      int n2 = stack.top();
      stack.pop();
      int n1 = stack.top();
      stack.pop();
      stack.push(n1 + n2);
    } else if (token == "-") {
      int n2 = stack.top();
      stack.pop();
      int n1 = stack.top();
      stack.pop();
      stack.push(n1 - n2);
    } else if (token == "*") {
      int n2 = stack.top();
      stack.pop();
      int n1 = stack.top();
      stack.pop();
      stack.push(n1 * n2);
    } else if (token == "/") {
      int n2 = stack.top();
      stack.pop();
      int n1 = stack.top();
      stack.pop();
      stack.push(n1 / n2);
    } else {
      stack.push(stoi(token));
    }
  }

  return stack.top();
}

